File: bindings/java/src/test/java/ua/dstucrypto/dstucore/StreamCipherTest.java
package ua.dstucrypto.dstucore;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertThrows;

/**
 * {@code crypto_stream} (Strumok-256 keystream) - <strong>no authentication</strong>: no
 * rejection category, since there is no tag to tamper with - {@link StreamCipher#decrypt} never
 * fails on tampered input, it silently returns different, wrong plaintext instead. Correctness:
 * round trip. Misuse: wrong-length key, truncated input.
 */
class StreamCipherTest {
    @Test
    void encryptDecryptRoundTrips() throws Exception {
        byte[] key = StreamCipher.keygen();
        byte[] sealed = StreamCipher.encrypt(key, "message".getBytes("UTF-8"));
        assertArrayEquals("message".getBytes("UTF-8"), StreamCipher.decrypt(key, sealed));
    }

    /**
     * Documents the no-integrity property explicitly, per this project's own precedent
     * ({@code hazmat::kalyna_xts}'s tampered-ciphertext test) - a deliberate design property, not
     * a missing rejection test.
     */
    @Test
    void tamperingIsNotDetectedButProducesWrongPlaintext() throws Exception {
        byte[] key = StreamCipher.keygen();
        byte[] sealed = StreamCipher.encrypt(key, "message".getBytes("UTF-8"));
        sealed[sealed.length - 1] ^= 1;
        byte[] garbage = StreamCipher.decrypt(key, sealed);
        assertFalse(java.util.Arrays.equals(garbage, "message".getBytes("UTF-8")));
    }

    @Test
    void wrongLengthKeyIsRejected() {
        byte[] tooShort = "too short".getBytes();
        assertThrows(IllegalArgumentException.class, () -> StreamCipher.encrypt(tooShort, "message".getBytes()));
    }

    @Test
    void truncatedSealedInputIsRejected() {
        byte[] key = StreamCipher.keygen();
        assertThrows(DstuException.class, () -> StreamCipher.decrypt(key, "short".getBytes()));
    }
}

File: bindings/go/dstu/streamcipher_test.go
package dstu

import (
	"bytes"
	"testing"
)

// StreamCipherKey (crypto_stream, Strumok-256 keystream) - no authentication: no rejection
// category, since Decrypt never fails on tampered input, it silently returns different, wrong
// plaintext instead. Correctness: round trip. Misuse: wrong-length key, truncated input.

func TestStreamCipherEncryptDecryptRoundTrips(t *testing.T) {
	key, _ := GenerateStreamCipherKey()
	defer key.Close()
	sealed, err := key.Encrypt([]byte("message"))
	if err != nil {
		t.Fatal(err)
	}
	got, err := key.Decrypt(sealed)
	if err != nil {
		t.Fatal(err)
	}
	if !bytes.Equal(got, []byte("message")) {
		t.Fatalf("got %q, want %q", got, "message")
	}
}

func TestStreamCipherTamperingIsNotDetectedButProducesWrongPlaintext(t *testing.T) {
	key, _ := GenerateStreamCipherKey()
	defer key.Close()
	sealed, _ := key.Encrypt([]byte("message"))
	sealed[len(sealed)-1] ^= 1
	garbage, err := key.Decrypt(sealed)
	if err != nil {
		t.Fatal(err)
	}
	if bytes.Equal(garbage, []byte("message")) {
		t.Fatal("expected tampered decryption to differ from the original plaintext")
	}
}

func TestStreamCipherWrongLengthKeyIsRejected(t *testing.T) {
	if _, err := StreamCipherKeyFromBytes(make([]byte, 10)); err == nil {
		t.Fatal("expected an error")
	}
}

func TestStreamCipherTruncatedSealedInputIsRejected(t *testing.T) {
	key, _ := GenerateStreamCipherKey()
	defer key.Close()
	if _, err := key.Decrypt([]byte("short")); err == nil {
		t.Fatal("expected an error")
	}
}
